### PR TITLE
[syslog] Eliminate heap allocations in log path

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Eliminates heap allocations in the syslog component's hot path by building the entire syslog packet incrementally in a single 508-byte stack buffer.

**Before:** Each log message caused 2 heap allocations:
1. `std::string timestamp` from `strftime()`
2. `std::string` from `str_sprintf()` for the packet

**After:** Zero heap allocations per log message:
- Single 508-byte stack buffer for the entire packet
- PRI, timestamp, hostname, tag, and message written incrementally with proper bounds checking
- Messages exceeding 508 bytes are safely truncated

**Key improvements:**
- Early return if PRI write fails (malformed packet prevention)
- Fallback to NILVALUE "-" if strftime fails (RFC 5424 compliance)
- Proper snprintf return value handling for all edge cases
- Comprehensive integration test with UDP listener

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Follow-up to #12588

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- N/A (no user-facing changes)

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No config changes - existing syslog configurations work unchanged
syslog:
  udp_id: syslog_udp
  time_id: sntp_time
  port: 514
  level: DEBUG
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: tests/integration/test_syslog.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        async with run_compiled(config), api_client_connected() as client:
+            # Verify device is running
+            device_info = await client.device_info()
+            assert device_info is not None
+            assert device_info.name == "syslog-test"
+
+            # Wait for syslog messages (ESPHome logs during startup)
+            try:
+                await receiver.wait_for_messages(timeout=10.0)
+            except TimeoutError:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/integration/test_syslog.py:167]: ruff is just going to change it back

File: esphome/components/syslog/esphome_syslog.cpp [hunk 2]
```diff
@@ -34,24 +34,40 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
+  }
+
+  // Write hostname, tag, and message
+  ret = snprintf(packet + offset, sizeof(packet) - offset, " %s %s: %.*s", App.get_name().c_str(), tag, (int) len,
+                 message);
+  if (ret > 0)
+    offset += ret;
+
+  if (offset > 0) {
+    this->parent_->send_packet(reinterpret_cast<const uint8_t *>(packet), std::min(offset, sizeof(packet) - 1));
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp]: 496c09b333091c6c8a948621046d0eda58deafb0

File: tests/integration/test_syslog.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        loop = asyncio.get_running_loop()
+        while True:
+            try:
+                data = await loop.sock_recv(sock, 4096)
+                if data:
+                    msg = data.decode("utf-8", errors="replace")
+                    receiver.on_message(msg)
+            except BlockingIOError:
+                await asyncio.sleep(0.01)
+            except Exception:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/integration/test_syslog.py:113]: this is a bit of a non-issue in the test

File: esphome/components/syslog/esphome_syslog.cpp [hunk 4]
```diff
@@ -34,24 +34,40 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
+    offset += now.strftime(packet + offset, sizeof(packet) - offset, "%b %e %H:%M:%S");
+  } else {
+    packet[offset++] = '-';
+  }
+
+  // Write hostname, tag, and message
+  ret = snprintf(packet + offset, sizeof(packet) - offset, " %s %s: %.*s", App.get_name().c_str(), tag, (int) len,
+                 message);
+  if (ret > 0)
+    offset += ret;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp]: 496c09b333091c6c8a948621046d0eda58deafb0

File: esphome/components/syslog/esphome_syslog.cpp [hunk 5]
```diff
@@ -34,24 +34,40 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
   size_t len = message_len;
   // remove color formatting
   if (this->strip_ && message[0] == 0x1B && len > 11) {
     message += 7;
     len -= 11;
   }
 
-  auto data = str_sprintf("<%d>%s %s %s: %.*s", pri, timestamp.c_str(), App.get_name().c_str(), tag, len, message);
-  this->parent_->send_packet((const uint8_t *) data.data(), data.size());
+  // Build syslog packet on stack - 508 is max UDP packet size
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp]: 496c09b333091c6c8a948621046d0eda58deafb0

File: esphome/components/syslog/esphome_syslog.cpp [hunk 6]
```diff
@@ -34,24 +34,40 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
+  int ret = snprintf(packet, sizeof(packet), "<%d>", pri);
+  if (ret > 0)
+    offset = ret;
+
+  // Write timestamp directly into packet (RFC 5424: use "-" if time not valid)
+  auto now = this->time_->now();
+  if (now.is_valid()) {
+    offset += now.strftime(packet + offset, sizeof(packet) - offset, "%b %e %H:%M:%S");
+  } else {
+    packet[offset++] = '-';
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp]: 496c09b333091c6c8a948621046d0eda58deafb0

File: esphome/components/syslog/esphome_syslog.cpp [hunk 7]
```diff
@@ -34,24 +34,40 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
-  auto data = str_sprintf("<%d>%s %s %s: %.*s", pri, timestamp.c_str(), App.get_name().c_str(), tag, len, message);
-  this->parent_->send_packet((const uint8_t *) data.data(), data.size());
+  // Build syslog packet on stack - 508 is max UDP packet size
+  char packet[508];
+  size_t offset = 0;
+
+  // Write PRI
+  int ret = snprintf(packet, sizeof(packet), "<%d>", pri);
+  if (ret > 0)
+    offset = ret;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp]: 496c09b333091c6c8a948621046d0eda58deafb0

File: tests/integration/test_syslog.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                )
+
+            # Verify message is truncated to max 508 bytes
+            assert len(trunc_msg) <= 508, f"Message exceeds 508 bytes: {len(trunc_msg)}"
+
+            # Verify the message starts correctly but is truncated (no "|END")
+            assert "START|" in trunc_msg, "Message should contain START marker"
+            assert "|END" not in trunc_msg, (
+                "Message should be truncated before END marker"
+            )
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/integration/test_syslog.py:260]: Its already well tested. This seems overkill

File: tests/integration/test_syslog.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+            assert log_service is not None, "log_long_message service not found"
+
+            # Call the service to trigger a long log message
+            await client.execute_service(log_service, {})
+
+            # Wait specifically for the truncation test message
+            try:
+                trunc_msg = await receiver.wait_for_pattern(r"trunctest.*START\|")
+            except TimeoutError:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/integration/test_syslog.py:248]: No, ruff will just remove it

File: esphome/components/syslog/esphome_syslog.cpp [hunk 10]
```diff
@@ -34,24 +34,47 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
+  char packet[508];
+  size_t offset = 0;
+  size_t remaining = sizeof(packet);
+
+  // Write PRI
+  int ret = snprintf(packet, remaining, "<%d>", pri);
+  if (ret > 0 && static_cast<size_t>(ret) < remaining) {
+    offset = ret;
+    remaining -= ret;
+  }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp:54]: a9c294bc0

File: esphome/components/syslog/esphome_syslog.cpp [hunk 11]
```diff
@@ -34,24 +34,47 @@ void Syslog::log_(const int level, const char *tag, const char *message, size_t
...
+
+  // Write timestamp directly into packet (RFC 5424: use "-" if time not valid)
+  auto now = this->time_->now();
+  if (now.is_valid()) {
+    size_t written = now.strftime(packet + offset, remaining, "%b %e %H:%M:%S");
+    offset += written;
+    remaining -= written;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/syslog/esphome_syslog.cpp]: a9c294bc0

**Review by clydebarrow — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks
